### ISSUE TITLE
forget: Document that time spans (week, month) are absolute

### ISSUE THREAD

**Issue by darkdragon-001** [restic/restic#2747]
Output of `restic version`
--------------------------
restic 0.9.6 compiled with go1.12.12 on linux/amd64

How did you run restic exactly?
-------------------------------

    restic -r /testrepo forget --keep-yearly 1 --keep-monthly 1 --keep-weekly 1 --keep-daily 1 --keep-last 1

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
local

Expected behavior
-----------------
Keep 1 year of snapshots.

Actual behavior
---------------
Only the last snapshot was retained.

Steps to reproduce the behavior
-------------------------------
See _How did you run restic exactly?_

Do you have any idea what may have caused this?
-----------------------------------------------
Design decision of using always the _last_ snapshot. This way, specifying an option only makes sense if the value ist >=2. It would be more intuitive to have values >=1 if an option should have an effect.

Do you have an idea how to solve the issue?
-------------------------------------------
For every interval, use earliest (instead of latest) snapshot.
`--keep-last` should default to 1 (this way the last snapshot will always be preserved when not overwritten). This also maintains backwards compatibility such that only _more_ and not _less_ snapshots are preserved for existing scripts after the change.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
It looks like restic will be a suitable tool for my backups (open source, cloud storage backends, secure).

**Comment by felix9** [comment]
This bothers me too. I'd like to make a PR to change this, but I'm wondering if this behavior needs a flag.

It seems to me that the current behavior is not useful, but I'm wary of making a change like this without giving people a chance to opt-in.

**Comment by rawtaz** [comment]
> Keep 1 year of snapshots.

This is not what `--keep-yearly` does. It keeps one snapshot per year, so e.g. if you have made a snapshot every day for 14 months, it will keep one snapshot that is 12 months/one year back in time from now.

> Only the last snapshot was retained.

When you write "last", do you mean the "most recent" or the "oldest" snapshot?

> `--keep-yearly 1 --keep-monthly 1 --keep-weekly 1 --keep-daily 1 --keep-last 1`

With this [retention policy](https://restic.readthedocs.io/en/latest/060_forget.html#removing-snapshots-according-to-a-policy), if you have made one snapshot every day for 14 months, you would get to keep:

- 1 snapshot from 12 months back in time.
- 1 snapshot from one month back in time.
- 1 snapshot from one week back in time.
- 1 snapshot from one day back in time.
- 1 snapshot that is the very latest snapshot.

So five snapshots in total (for each combination of `--group-by` you have specified). Is this not what you are seeing?

In general I'm not sure I understand what you are saying is the problem and what you are suggesting instead.

**Comment by rawtaz** [comment]
Okay, I get it now, I think. You want the `--keep-yearly` to mean "keep every snapshot for an entire year". Basically you want a `keep-year` rather than `--keep-yearly`.

Depending on how many snapshots you do every day, this can have a lot of unexpected snapshots preservations though. Consider if you make one snapshot every hour for twelve hours every day, and `--keep-yearly` meant what I wrote above. Then you'd keep 4380 snapshots, and it's unlikely that you want to keep every hour one year back.

The way it currently works is more specific, keeps only what you ask for, without keeping more than that, and you are still perfectly able to specify what you (if I'm right) want. Unless I'm missing something here, this should not be changed from how it's currently designed, as that would cause it to be less flexible and do more than what the option suggests.

**Comment by fd0** [comment]
I think you're looking for the following option:
```
      --keep-within duration    keep snapshots that are newer than duration (eg. 1y5m7d2h) relative to the latest snapshot
```

Just run `restic forget --keep-within 1y` to retain all snapshots made within the last year of the latest snapshot.

Is that what you need?

**Comment by fd0** [comment]
I'm closing this issue for now, please feel free to add further comments. Thanks a lot!

**Comment by felix9** [comment]
no, let me try to explain what I want.

Let's say, in general, I want to always have a backup from about a year ago.

I start doing backups every day from 2021-01-30, and also run forget every day.

When it's 2022-01-30, I want to have at least
- The daily backup, 2022-01-30
- The backup from 2021-01-30

and ideally not much more.

What forget policy should I use to get that?

If I say --keep-yearly 1, then 2021-01-30 will disappear as soon as I have 2021-01-31
If I say --keep-yearly 2, the same thing happens.
If I say --keep-yearly 2 --keep-monthly 13, the same thing happens.

In order to keep the 2021-01-30 snapshot,
I need to specify a combination of options that will keep it around for a year, 
at which time it will be kept by --keep-yearly 2.

So to keep 2021-01-30 until at least 2022-01-30,
 which satisfies "keep at least one backup from about a year ago",
 I need to specify something like --keep-yearly 2 --keep-monthly 13 --keep-daily 32,
 which seems quite counter-intuitive to me,
 and also keeps many more snapshots than I really want.

What I want is some less confusing way to say "keep at least one backup from about a year ago".

Algorithmically, to keep "at least one backup from the last n years":
 for each n+1 one-year intervals before now,
 keep the earliest backup that exists in that interval.

**Comment by fd0** [comment]
Thanks for the explanation!

> When it's 2022-01-30, I want to have at least
> * The daily backup, 2022-01-30
> * The backup from 2021-01-30
> and ideally not much more.
> 
> What forget policy should I use to get that?

I'd use `restic forget --keep-daily 14 --keep-monthly 13`, which would get you very close:

 * It'd keep the daily backup 2022-01-30 (as a daily backup due to `--keep-daily 14`)
 * It'd keep the backup on 2022-01-31 (as the latest in that month, due to `--keep-monthly 13`)

This should satisfy your policy of keep a backup from about a year ago.

> If I say --keep-yearly 1, then 2021-01-30 will disappear as soon as I have 2021-01-31

That's not quite right, it'll be removed even earlier: with `--keep-yearly 1` (standalone without any other options), you instruct restic to keep the latest backup for the latest year in which a snapshot was made. Since each new daily backup is sufficient to fulfill this condition, restic will only keep the newest snapshot and remove everything else. So you always only have exactly one snapshot in the repo: the latest daily backup. On 2022-01-30 you'll end up with the snapshot from 2022-01-30.

I agree that most of the settings do not make much sense when set to the value 1.

> If I say --keep-yearly 2, the same thing happens.

Also not quite right: with `--keep-yearly 2` (standalone without any other options), you instruct restic to keep the latest backup for the two latest years in which a snapshot was made. Throughout 2021, it'll behave like `--keep-yearly 1` (because there's only one year in which snapshots were made). After the daily backup on 2022-01-01, it'll keep that daily backup as well as the snapshot from 2021-12-31. On 2022-01-30, you'll have two snapshots in the repo: the one from 2021-12-31 as well as the latest daily backup from 2022-01-30.

> If I say --keep-yearly 2 --keep-monthly 13, the same thing happens.

In this case, throughout 2021, restic will keep the latest daily snapshot as well as the latest backup within each month for the latest 13 months. So on 2021-12-15 you'll have 12 snapshots:
 * 2021-01-31
 * 2021-02-28
 * 2021-03-31
 * [...]
 * 2021-12-15

On 2022-01-30 you'll have 13 snapshots:
 * 2021-01-31
 * 2021-02-28
 * 2021-03-31
 * [...]
 * 2021-12-31
 * 2022-01-30

Here, `--keep-yearly 2` does not matter (yet), because the monthly snapshots also count as yearly snapshots, and you only have two years.

> Algorithmically, to keep "at least one backup from the last n years":
> for each n+1 one-year intervals before now,
> keep the earliest backup that exists in that interval.

I can see what you're trying to do, but (as you already said) the design decision was made when I implemented the retention policies to always keep the newest, most recent snapshot when several snapshots match. The idea behind it was that (in my opinion at least) data tends to be less relevant the older it is, so when restic needs to chose between several snapshots it could keep to satisfy a policy, it'll choose the most recent one.

The expiration policy settings are already complicated enough, which surprised me a lot after I implemented them and then had to explain it to other people. I'm very hesitant to add further, even more complicated policies or settings for `restic forget`.

You have several options now:
 * Live with the fact that you have to keep a few more snapshots around (caused by `--keep-monthly 13` or something similar)
 * Add a tag to snapshots that you'd like to keep and pass `--keep-tag foo`, add/remove such tags as needed, for example with a script parsing `restic snapshots --json`
 * Implement your own expiration policy by parsing the output of `restic snapshots --json` and pass the list of snaphost to remove to `restic forget`

Thanks for understanding!

**Comment by felix9** [comment]
> I'd use `restic forget --keep-daily 14 --keep-monthly 13`, which would get you very close:
> 
> * It'd keep the daily backup 2022-01-30 (as a daily backup due to `--keep-daily 14`)
> * It'd keep the backup on 2022-01-31 (as the latest in that month, due to `--keep-monthly 13`)

But that doesn't work if I run forget daily, because I'll only ever have the most recent 14 days of backups. I'll never have a backup that will match `--keep-monthly 13`, other than the current month.

It does work if I run forget monthly.

It feels very strange to me that the current forget behavior has different behavior depending on how often you run it.

The algorithm I suggested (keeping the oldest snapshot in a timeslice) doesn't have this problem. It behaves pretty much the same no matter how often you run it.

**Comment by felix9** [comment]
here's the algorithm that I want:

for a group of snapshots,
always keep the most recent
keep additional recent if --keep-last > 1

for each time-based `--keep-T N` option,
divide the time before now into N+1 segments are T units long,
keep the oldest snapshot that exists in each segment.

I think this is generally how people expect --keep options to work.
it doesn't have the counter-intuitive behavior of the current algorithm.
it will always keep at least as many snapshots as the current algorithm.
it might keep 1 more snapshot per --keep-T option, but it won't keep more than that.
it will preserve more history than the current algorithm.
it will tend to keep snapshots that are more recent than the ones kept by the current algorithm. (this is perhaps counter-intuitive, but it's easy to demonstrate: say you have 4 weeks of daily backups and you say --keep-weekly 2. the current algorithm will keep the one today and the one from 7 days ago. my suggested algorithm will keep the one today, the one from 6 days ago, and the one from 13 days ago)

implementation plan:

I'd add a flag named something like `--keep-strategy`.
the existing algorithm would be `--keep-strategy classic`.
the new algorithm would be `--keep-strategy archive`.

the default wouldn't change.
(though it might be useful to warn or error if there isn't an explicit --keep-strategy)

**Comment by fd0** [comment]
> But that doesn't work if I run forget daily, because I'll only ever have the most recent 14 days of backups. I'll never have a backup that will match --keep-monthly 13, other than the current month.

Huh, it feels to me that we still have a misunderstanding here. If you always run `forget` with the same parameters, its behavior does not change, independently on how often you run it. It is idempotent.

I'll show you an example. You can follow along by first building restic with the `debug` tag (so we can use an insecure KDF algorithm which will allow us to make many backups quickly:

    $ git clone https://github.com/restic/restic
    $ cd restic/cmd/restic
    $ go build -tags debug

Then initialize a new repository:

    $ export RESTIC_REPOSITORY=/tmp/repo RESTIC_PASSWORD=foo
    $ ./restic init --insecure-kdf

Then make 400 daily backups (with a simulated timestamp) of `main.go` and run `restic forget --keep-daily 14 --keep-monthly 13` afterwards:

    for i in {0..400}; do
        timestamp=$(date +"%Y-%m-%d 12:00:00" -d "+ ${i} days")
        echo "=========================================================="
        echo "run backup on ${timestamp}"
        ./restic backup --quiet --time "${timestamp}" main.go
        ./restic forget --compact --keep-daily 14 --keep-monthly 13
    done

The log can be found here: https://gist.github.com/fd0/16eb15698fcf5bad2a6a843acf3a5800

[At the end](https://gist.github.com/fd0/16eb15698fcf5bad2a6a843acf3a5800#file-log-L14864-L14894), you'll have the following snapshots left:

    keep 26 snapshots:
    ID        Time                 Host    Tags
    ---------------------------------------------
    33c54046  2021-01-31 12:00:00  mopped
    61beb7cd  2021-02-28 12:00:00  mopped
    490153a3  2021-03-31 12:00:00  mopped
    dfebd319  2021-04-30 12:00:00  mopped
    1d1ce948  2021-05-31 12:00:00  mopped
    9529f8ca  2021-06-30 12:00:00  mopped
    2c9f66f1  2021-07-31 12:00:00  mopped
    d77abab4  2021-08-31 12:00:00  mopped
    c009cda9  2021-09-30 12:00:00  mopped
    f823e805  2021-10-31 12:00:00  mopped
    99cf1075  2021-11-30 12:00:00  mopped
    efa08558  2021-12-31 12:00:00  mopped
    6186007c  2022-01-03 12:00:00  mopped
    8025af27  2022-01-04 12:00:00  mopped
    b7089acc  2022-01-05 12:00:00  mopped
    56eb92cd  2022-01-06 12:00:00  mopped
    161d15ce  2022-01-07 12:00:00  mopped
    684a86a9  2022-01-08 12:00:00  mopped
    e139f8cc  2022-01-09 12:00:00  mopped
    2a6af877  2022-01-10 12:00:00  mopped
    52508f1c  2022-01-11 12:00:00  mopped
    fed04645  2022-01-12 12:00:00  mopped
    0ee3287d  2022-01-13 12:00:00  mopped
    0a98f5f9  2022-01-14 12:00:00  mopped
    11bb53b3  2022-01-15 12:00:00  mopped
    3bef0251  2022-01-16 12:00:00  mopped
    ---------------------------------------------
    26 snapshots

Then do the same again but run forget only once at the very end (after 400 days of backups):

    $ for i in {0..400}; do
        timestamp=$(date +"%Y-%m-%d 12:00:00" -d "+ ${i} days")
        echo "=========================================================="
        echo "run backup on ${timestamp}"
        ./restic backup --quiet --time "${timestamp}" main.go
    done

    ./restic forget --compact --keep-daily 14 --keep-monthly 13

The log is here: https://gist.github.com/504a431b6f597b5681e6b3f0c2f1869a

[At the end](https://gist.github.com/fd0/504a431b6f597b5681e6b3f0c2f1869a#file-log2-L429-L458), you'll have the same snapshots:

    ID        Time                 Host    Tags
    ---------------------------------------------
    58acc8b6  2021-01-31 12:00:00  mopped
    84986195  2021-02-28 12:00:00  mopped
    adda4aac  2021-03-31 12:00:00  mopped
    ac8c789f  2021-04-30 12:00:00  mopped
    4f334cf6  2021-05-31 12:00:00  mopped
    ea23595e  2021-06-30 12:00:00  mopped
    a9f38ac0  2021-07-31 12:00:00  mopped
    55c025a3  2021-08-31 12:00:00  mopped
    4d0c214a  2021-09-30 12:00:00  mopped
    3da165f9  2021-10-31 12:00:00  mopped
    8a9ac25a  2021-11-30 12:00:00  mopped
    5edb8282  2021-12-31 12:00:00  mopped
    553288cc  2022-01-03 12:00:00  mopped
    541bb3d5  2022-01-04 12:00:00  mopped
    cff88fb0  2022-01-05 12:00:00  mopped
    c3889aa9  2022-01-06 12:00:00  mopped
    f0904222  2022-01-07 12:00:00  mopped
    84287ff3  2022-01-08 12:00:00  mopped
    0319c51a  2022-01-09 12:00:00  mopped
    8eb018ea  2022-01-10 12:00:00  mopped
    463848c5  2022-01-11 12:00:00  mopped
    3dc4a64b  2022-01-12 12:00:00  mopped
    0f677b70  2022-01-13 12:00:00  mopped
    8135460a  2022-01-14 12:00:00  mopped
    e7081ad6  2022-01-15 12:00:00  mopped
    124ef82e  2022-01-16 12:00:00  mopped
    ---------------------------------------------
    26 snapshots

> it will tend to keep snapshots that are more recent than the ones kept by the current algorithm. (this is perhaps counter-intuitive, but it's easy to demonstrate: say you have 4 weeks of daily backups and you say --keep-weekly 2. the current algorithm will keep the one today and the one from 7 days ago.

Almost: for the current algorithm, it depends on when the last snapshot was made, because it considers "week" as "calendar week", which starts with a Monday and end with a Sunday, not "week" as "a sequence of seven days". It's the same with "month": "calendar month, e.g. January" vs. "the last 30 days"(?).

> my suggested algorithm will keep the one today, the one from 6 days ago, and the one from 13 days ago)

So, it'd be approximately what `--keep-weekly 3` currently does, right?

> here's the algorithm that I want:
>
> for a group of snapshots,
> always keep the most recent
> keep additional recent if --keep-last > 1

This part you can already have by setting `--keep-last` to something ≥ 1.

> for each time-based --keep-T N option,
> divide the time before now into N+1 segments are T units long,
> keep the oldest snapshot that exists in each segment.

Depending on how you make backups, the outcome (the list of snapshots to be kept) is very similar to what you get with the current algorithm by passing your `N+1` to the `--keep-T` option. Instead of the first backup of a month, it'll keep the last one of the previous month, which (again, depending on your use case) gives you a restore point in time that's pretty similar to what your requirements were.

Other issues that I can see with this simplified pseudo-code algorithm:
 * How is `now` defined? For the current algorithm it's the timestamp of the most recently made snapshot (on purpose).
 * Is a `week` defined as the last seven days? How can the algorithm work with a sliding window?
 * How long is a `month`? 30 days? 31 days?

I'm not convinced that adding a (slightly different) retention policy algorithm is worth the complexity.

**Comment by felix9** [comment]
Yeah, I think we're still talking past each other :)

Here's a simplified example.

Say I start doing daily backups on day 1.

I want to keep about 3 daily backups and 3 weekly backups.

On day 21 I'd like to have backups from day 21, 20, 19, 14, 7. (Or something similar)

It seems like the right way to get that is something like `--keep-daily 3 --keep-weekly 3`

So each day, after running backup and then forget, this is what I have:

day 1: 1
day 2: 2, 1
day 3: 3, 2, 1
day 4: 4, 3, 2
day 5: 5, 4, 3
day 6: 6, 5, 4
day 7: 7, 6, 5

etc.

In order for the --keep-weekly 3 to work, I need some other options that will keep some backup for at least a week.

Alternatively, I could reduce the frequency of forget: if I run forget weekly instead of daily, it will work how I want. Well, it will probably work how I want. Since scheduling is never exact, it's hard to depend on that.

This is what I find extremely counter-intuitive about the current algorithm. It's hard to plan how to configure a system to reliably get a result like "keep 3 daily, keep 3 weekly", which is something that seems like it should be simple.

My proposed alternate algorithm will do that easily.

**Comment by rawtaz** [comment]
> I want to keep about 3 daily backups and 3 weekly backups.
> 
> On day 21 I'd like to have backups from day 21, 20, 19, 14, 7. (Or something similar)

Before I ask my question below, I have to point out that using terms as "about" and "something similar" does not work when discussing an exact algorithm. Restic keep policy doesn't do "something similar" to, it does "exactly" something :)

That said, can you please explain why you want to keep a few consecutive days a long way back in time? What's the use case for this? The common use cases are usually based around wanting more frequent snapshots restorable close to current time, and the older snapshots are keeping fewer of them (as in, if you want to restore a file you lost the last week you want to be able to go down to days or even hours, while if you're looking to restore a file you lost sometime during the last year, you're more interested in going down to a resolution of e.g. months). If we extend your example, why on earth would you want to keep e.g. a week of daily snapshots from six months back in time, while you don't keep a week of snapshots for the last week? I fail to see how any of this makes sense.

**Comment by zroug** [comment]
I think the issue that @felix9 has is that when you always keep the most recent snapshot of each wanted interval, then no snapshot ever gets old because the old snapshot gets deleted before it falls into a wanted interval. Example: Policy is `--keep-weekly 3`. Backup and forget every day. Expected behavior: After three weeks you have a snapshot for each week. Current behavior:

| Day  | Snapshots Left |
| - | - |
| 1 | s1 |
| 2 | s2 |
| ... | ... |
| 30 | s30 |

there is always only one snapshot left after forget.

With @felix9 proposed algorithm (with this algorithm the most recent snapshot must always be kept) it looks like this:

| Day  | Snapshots Left |
| - | - |
| 1 | s1 |
| 2 | s2, s1 |
| 3 | s3, s1 |
| 4 | s4, s1 |
| 5 | s5, s1 |
| 6 | s6, s1 |
| 7 | s7, s1 |
| - | - |
| 8 | s8, s1 |
| 9 | s9, s8, s1 |
| 10 | s10, s8, s1 |
| 11 | s11, s8, s1 |
| 12 | s12, s8, s1 |
| 13 | s13, s8, s1 |
| 14 | s14, s8, s1 |
| - | - |
| 15 | s15, s8, s1 |
| 16 | s16, s15, s8, s1 |
| 17 | s17, s15, s8, s1 |
| 18 | s18, s15, s8, s1 |
| 19 | s19, s15, s8, s1 |
| 20 | s20, s15, s8, s1 |
| 21 | s21, s15, s8, s1 |
| - | - |
| 22 | s22, s15, s8 |
| 23 | s23, s22, s15, s8 |
| 24 | s24, s22, s15, s8 |
| 25 | s25, s22, s15, s8 |
| 26 | s26, s22, s15, s8 |
| 27 | s27, s22, s15, s8 |
| 28 | s28, s22, s15, s8 |

I think this is exactly the expected behavior. I also do not think that you want the most recent data necessarily. When you can't use the most recent snapshot, it is unclear how old the snapshot should be, therefore everything that matches the forget policy should be fine.

For my backup script I had to implement a custom policy because of this issue. It is not intuitive at first to keep the oldest snapshot instead of the most recent but I think it is the correct thing to do (as long as you also keep the latest).

Also @felix9 didn't mean to keep the 3 consecutive snapshots for a period long way back in time. His policy would still use the most recent days but for each given interval (an interval being a week, day, ...) he keeps the oldest snapshot that still is in the interval.

**Comment by fd0** [comment]
Very interesting, thanks for your patience and taking the time to explain it again! I'm reopening this issue. For the next step, I'd like to define what the problem with the current algorithm is. Then we can discuss what we should do about it, maybe there are other solutions than 

I think there's a valid issue that you've pointed out: the way restic's retention policies are implemented, you can build policies which are not intuitive and do not make sense. I wasn't aware that such policies exist. In these cases, it is indeed relevant when `restic forget` is run.

You gave a good example: With daily `restic backup` followed by `restic forget --keep-weekly 3`, it will always only keep the latest daily snapshot.

That's the core issue with the current implementation as far as I understood it now. Is that correct? Is there anything else missing?

**Comment by zroug** [comment]
Yes I think that is the core issue. In fact I think that for any combination of the time based policies (`-H -d -w -m -y`), if only you execute backup and forget often enough, only the latest snapshot is kept.

The reason is that snapshots never have the chance to get old enough to reach the next time interval. Keeping the oldest snapshot for each wanted interval is one relatively simple solution to this problem. There might be other solutions but I don't know about them.

**Comment by fd0** [comment]
Sigh, this is surprisingly complicated. It's been a long time since I worked on the retention policy algorithm.

Did you actually try running restic with `forget --keep-weekly 3`?

Because I just did, and the algorithm works as expected. Initialize a repo:

    cd cmd/restic
    go build --tags debug
    export RESTIC_REPOSITORY=/tmp/repo RESTIC_PASSWORD=foo
    ./restic init --insecure-kdf

Then simulate 40 daily snapshots:

    for i in {0..40}; do
        timestamp=$(date +"%Y-%m-%d 12:00:00" -d "+ ${i} days")
        echo "=========================================================="
        echo "run backup on ${timestamp}"
        ./restic backup --quiet --time "${timestamp}" main.go
        ./restic forget --keep-weekly 3
    done

The log is here: https://gist.github.com/100ff7f26f816caa9531820a5e45e8de

Afterwards, the list of snapshots looks like this, just as the policy specified:

    ID        Time                 Host        Tags        Paths
    -------------------------------------------------------------------------------------------------------------------------------
    ed1418b8  2021-01-10 12:00:00  mopped                  /home/fd0/shared/work/go/src/github.com/restic/restic/cmd/restic/main.go
    ae0607be  2021-01-17 12:00:00  mopped                  /home/fd0/shared/work/go/src/github.com/restic/restic/cmd/restic/main.go
    6d5d1008  2021-01-23 12:00:00  mopped                  /home/fd0/shared/work/go/src/github.com/restic/restic/cmd/restic/main.go
    -------------------------------------------------------------------------------------------------------------------------------
    3 snapshots

It's the same when `restic forget` is run weekly, or even only at the end.

Am I missing something?

**Comment by zroug** [comment]
I am very sorry, I did not test that. As I have said, when implementing my own retention policy, I had the very same issue and according to the documentation restic should have it to. I thought there was a simple miscommunication happening and wanted to help to clear that up but I did not test the issue itself enough. Apparently restic does not suffer from what I thought it does.

But there might be a documentation issue. According to your log (https://gist.github.com/fd0/100ff7f26f816caa9531820a5e45e8de#file-log-L126-L134) on 2020-12-21 the snapshots 2020-12-20 and 2020-12-21 are kept. The documentation says

> `--keep-weekly n` for the last `n` weeks which have one or more snapshots, only keep the last one for that week.

That sounds like 2020-12-20 should be removed and it is unclear why it is kept. (Of course it is good that it is kept, otherwise this issue would be correct.) I think the documentation might be the reason that this issue was opened in the first place. At least it was the reason why I thought it was plausible.

Sorry again that I did not test that beforehand.

**Comment by fd0** [comment]
@zroug don't worry, you tried to help and got me thinking again, so thanks!

> In order for the --keep-weekly 3 to work, I need some other options that will keep some backup for at least a week.

@felix9 I think what you wrote isn't correct and the current algorithm works, please have a look at my test case and let me know if I'm missing anything. :)

I'm now convinced that the underlying issue is the documentation: It does not talk about what a "week" is, and that's also missing from your example. After the first backup on a Monday, restic will see two weekly backups: The one from Sunday is the last week, and the one from Monday is in the current week.

It's the same for daily (transition: 23:59 -> 00:00), monthly (transition 31st -> 1st) and yearly (31 December -> 1 January) snapshots.

I'm keeping this as a documentation issue. Is anyone up for trying to improve the docs? :)

**Comment by rawtaz** [comment]
It's a bit weird though. Noone with their sane mind would expect an "hour" to mean anything than xx:00 -> xx:59, i.e. the natural boundary instead of an hour starting. This doesn't need mentioning in the docs.

So why then would someone presume that days, weeks, months and years would work differently than being on natural boundaries, too? Doesn't really make sense.

**Comment by zroug** [comment]
For me that PR makes it perfectly clear. I don't think its unreasonable to think of 9:30 to 10:30 when you speak about the last hour at 10:30. (Perhaps the strength of this assumption is different in different countries, I don't know.) Once you are in that mindset it is easy to think of the last three hours as 7:30 to 10:30 and simply continue this thinking for the larger time frames. Sometimes its difficult to think outside the box when you have the wrong assumption in the first place. But just a little push in the right direction is enough (for me that was when I realized that a snapshot is kept at the boundary from Sunday to Monday and now the docs will do that).

**Comment by rawtaz** [comment]
> For me that PR makes it perfectly clear.

Thanks! I'll wait for others to chime in as well, just in case something is still unclear (but please note that we want to keep the text from exploding into a large paragraph :D).

> I don't think its unreasonable to think of 9:30 to 10:30 when you speak about the last hour at 10:30.

Given that this is the first time I've seen this confusion in many many years of being in the restic community, I'd say it's extremely rare or unlikely to happen.

> Sometimes its difficult to think outside the box when you have the wrong assumption in the first place.

This is probably very true.

**Comment by felix9** [comment]
ah, ok, that makes sense now. I think my confusion was `--keep-weekly` doesn't seem to me like it means pinned to an absolute partitioning, since the boundaries of a week are arbitrary. (yes, hour is less arbitrary. but month and year are dependent on calendar, etc.)

Thanks for tolerating my confusion.

I still have one objection to latest-in-interval, instead of earliest-in-interval :)

I think in general, keeping earliest-in-interval tends to keep more deltas than latest-in-interval, which means I can reconstruct more past states.

consider, I start taking daily backups with a `--keep-weekly 3` policy, and then delete a bunch of files that I don't need, reasoning that if I made a mistake and really do need them, I can restore them from my backups.

but forget's behavior means that as soon as I take another backup, it's likely to remove the snapshot that has the deleted files.

basically, with current behavior, it's easy to recover from mistakes made since the latest backup, but it's less easy to recover from mistakes made before the latest backup, without careful consideration of the forget options and backup schedule.

I think 'keep earliest in interval' has more intuitive behavior for archival/recovery purposes.

If adding this option is really a non-starter, then I'll look at documenting 'how to do x' with existing behavior.

**Comment by rawtaz** [comment]
Just adding +1 to your policies will keep an additional snapshot of that policy type, doesn't this cover your concern?

**Comment by fd0** [comment]
> ah, ok, that makes sense now. I think my confusion was --keep-weekly doesn't seem to me like it means pinned to an absolute partitioning, since the boundaries of a week are arbitrary. (yes, hour is less arbitrary. but month and year are dependent on calendar, etc.)
>
>Thanks for tolerating my confusion.

Thanks for the feedback! It showed us that there's something missing from the docs, so the docs got better. That's a good result :)

> If adding this option is really a non-starter, then I'll look at documenting 'how to do x' with existing behavior.

Unfortunately, in my opinion, such a change is still not worth it. Changing the default behavior is bad for users, adding a second strategy is not worth the complexity. So, yes, if you think the docs can still be improved, then let's do that.

> Just adding +1 to your policies will keep an additional snapshot of that policy type, doesn't this cover your concern?

There are still corner cases I can think of in which "keep least recent snapshot" may keep more snapshots, and in these situations adding 1 to the retention policy won't help. But in general I think the retention policy system restic currently has is expressive enough.

**Comment by darkdragon-001** [comment]
I finally found some time to read through your very helpful comments. Thanks for that! Especially @fd0 scripts to test policies out easily is really helpful!

> I think in general, keeping earliest-in-interval tends to keep more deltas than latest-in-interval, which means I can reconstruct more past states.

This is still my opinion as well. Especially that `--keep-X 1` doesn't add anything if another option for a smaller interval is used. This is quite counter-intuitive for me (why is it possible in the first place)! Especially that a dedicated `--keep-last N` option exists! When it's unfeasible to change the current behavior for backwards-compatibility, at least a warning should be printed when using `--keep-X 1`!

Adding `+1` to all my intervals leads to the desired behavior for my use case.

**Comment by fd0** [comment]
> a warning should be printed when using --keep-X 1!

That's a good idea! It shouldn't be that hard to check for these policies so we can print a warning if restic is run interactively. It should only print such a warning if the check is sure that an option does not add anything (like, `restic forget --keep-daily 5 --keep-weekly 1` should be warned about that the `--keep-weekly 1` is already covered).

Anybody up for a PR?

**Comment by darkdragon-001** [comment]
For simplicity, I would _always_ warn when using `--keep-[hourly|daily|weekly|monthly|yearly] 1` as this behaves likes `--keep-last 1` which one can assume is not intended (otherwise people would use `--keep-last 1` directly).

**Comment by darkdragon-001** [comment]
@fd0 can you plese reopen as this isn't resolved? Or should I create a new issue?

**Comment by MichaelEischer** [comment]
@darkdragon-001 I think it's better to open a new issue. It took me quite a while to find out that you're only referring to the last few comments in this issue?
